### PR TITLE
Add the openfaas-fn namespace to function store Custom Resources

### LINKED ISSUE #892: Add namespace to generated functions
Running the generate command is not adding the function(s) namespace.

## Expected Behaviour
The following command

```
faas-cli generate \
  --from-store nodeinfo \
  > nodeinfo-store.yaml
```

Should give this

```
---
apiVersion: openfaas.com/v1
kind: Function
metadata:
  name: nodeinfo
  namespace: openfaas-fn
spec:
  name: nodeinfo
  image: ghcr.io/openfaas/nodeinfo:latest
  labels: {}
```

Notice the namespace.

## Current Behaviour
The namespace is not added, instead it gives this

```
---
apiVersion: openfaas.com/v1
kind: Function
metadata:
  name: nodeinfo
spec:
  name: nodeinfo
  image: ghcr.io/openfaas/nodeinfo:latest
  labels: {}
  annotations: {}

```

## Are you a GitHub Sponsor (Yes/No?)
Yes

Check at: https://github.com/sponsors/openfaas
- [x] Yes
- [ ] No

## List All Possible Solutions and Workarounds
The current workaround is to manually add the namespace to the generated file, or if using inline with kubectl, passing the flag `--namespace=openfaas=fn`.

## Which Solution Do You Recommend?
If the namespace is not provided in the function spec, it should use the default - openfaas-fn. If the function spec specifies a particular namespace, it should use that one.

## Steps to Reproduce (for bugs)
Taken from [here](https://www.openfaas.com/blog/manage-functions-with-kubectl/).
```
faas-cli generate \
  --from-store nodeinfo \
  > nodeinfo-store.yaml
```

## Context
Attempting to deploy to a cluster without access to the gateway.

## Your Environment

* FaaS-CLI version ( Full output from: `faas-cli version` ):

CLI:
 commit:  77ad215bcc6291dbf72c73caffe4d76aa2bb6fb1
 version: 0.13.11
Handling connection for 8080

Gateway
 uri:     http://127.0.0.1:8080
 version: 0.20.12
 sha:     a6dbb4cd0285f6dbc0bc3f43f72ceacdbdf6f227

Provider
 name:          openfaas-operator
 orchestration: kubernetes
 version:       0.13.2 
 sha:           6caf24ba21f7af171e2ade40daa033b05584b695

* Docker version ( Full output from: `docker version` ):

* Are you using Docker Swarm (FaaS-swarm ) or Kubernetes (FaaS-netes)?
FaaS-netes

* Operating System and version (e.g. Linux, Windows, MacOS):

* Link to your project or a code example to reproduce issue:

### PR DESCRIPTION
**Pull request by yankeexe**
Add `defaultFunctionNamespace` to avoid empty namespace on Kubernetes CRD YAML generation.

Resolves #892

Signed-off-by: Yankee Maharjan <yankee.exe@gmail.com>

## Description

Generates CRD YAML file with default namespace (`openfaas-fn`) if not provided.

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?

Generate YAML config with default namespace: 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
This change looks fine to me. Thanks for working on it so quickly.

For what it's worth, functions generate with a default namespace already, this seems to have been limited just to the store functions.

With:

```yaml
version: 1.0
provider:
  name: openfaas
  gateway: http://127.0.0.1:8080
functions:
  tester:
    lang: go
    handler: ./tester
    image: tester:latest
    annotations:
      com.openfaas.health.http.path: "/healthz"
```

```
faas-cli generate -f tester.yml 
---
apiVersion: openfaas.com/v1
kind: Function
metadata:
  name: tester
  namespace: openfaas-fn
spec:
  name: tester
  image: tester:latest
```

**Comment by alexellis** [comment]
@yankeexe can you check that the generate `--from-store` command takes a custom namespace flag also?

I suspect that it may not, so should do as part of the PR. If we are going to support a hard-coded namespace of openfaas-fn, then we need to offer an alternative.

This might also cause confusion when people pass `--namespace` whilst having a namespace configured within the YAML file.

Pinging @rgee0 who's helped with flag priority recently.

**Comment by alexellis** [comment]
/set title: Add the openfaas-fn namespace to function store Custom Resources

**Comment by rgee0** [comment]
Check #869 and #872 for additional context.

**Comment by yankeexe** [comment]
> For what it's worth, functions generate with a default namespace already, this seems to have been limited just to the store functions.
> ```
> faas-cli generate -f tester.yml 
> ---
> apiVersion: openfaas.com/v1
> kind: Function
> metadata:
>   name: tester
>   namespace: openfaas-fn
> spec:
>   name: tester
>   image: tester:latest
> ```

@alexellis
I am posting these results from `faas-cli` version **0.13.11**. Can you confirm these are the desired behavior?

Defining a default namespace for function store custom resource will fix the issue, but wanted to confirm other behaviors are as expected, and If not, I can include a fix on this PR. 

 For me there is still no namespace with the YAML file you provided: 

<details>
<summary> <b>No default namespace in CR with provided `tester.yml` </b></summary> 

</details>

> @yankeexe can you check that the generate --from-store command takes a custom namespace flag also?

It does take the namespace flag. 

<details>
<summary> <b>User defined namespace through flags with `--from-store` </b></summary> 

</details>


> This might also cause confusion when people pass --namespace whilst having a namespace configured within the YAML file.

Even with a namespace defined on the function, the generated custom resource doesn't take it into account.  Passing a  `--namespace` flag works fine.

<details>
<summary><b>With namespace on YAML, no namespace on Custom Resource</b></summary>

</details>

**Comment by goncalo-oliveira** [comment]
@alexellis here's an example without a store function.

```
$ faas-cli version
CLI:
 commit:  77ad215bcc6291dbf72c73caffe4d76aa2bb6fb1
 version: 0.13.11

$ faas-cli new --lang go --handler src hello

$ cat hello.yml
version: 1.0
provider:
  name: openfaas
  gateway: http://127.0.0.1:8080
functions:
  hello:
    lang: go
    handler: ./src
    image: hello:latest

$ faas-cli generate -f hello.yml 
---
apiVersion: openfaas.com/v1
kind: Function
metadata:
  name: hello
spec:
  name: hello
  image: hello:latest

$ nano hello.yml
$ cat hello.yml
version: 1.0
provider:
  name: openfaas
  gateway: http://127.0.0.1:8080
functions:
  hello:
    lang: go
    handler: ./src
    namespace: openfaas-fn
    image: hello:latest

$ faas-cli generate -f hello.yml
---
apiVersion: openfaas.com/v1
kind: Function
metadata:
  name: hello
spec:
  name: hello
  image: hello:latest

$ faas-cli generate --namespace openfaas-fn -f hello.yml 
---
apiVersion: openfaas.com/v1
kind: Function
metadata:
  name: hello
  namespace: openfaas-fn
spec:
  name: hello
  image: hello:latest
```

Hope this helps.

**Comment by alexellis** [comment]
I will check my faas-cli version. There's a chance that my copy is from before Richard's change. In which case @goncalo-oliveira is right about both parts having regression (sorry)

Yankee's patch reverts the change made by Richard, so we need to figure out how to make this only apply to the generate method.

**Comment by alexellis** [comment]
@rgee0 thanks for those two links. Please could you review this PR?

File: commands/generate.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ import (
 )
 
 const (
-	defaultFunctionNamespace = ""
+	defaultFunctionNamespace = "openfaas-fn"
```
Comment (alexellis) [inline commands/generate.go:24]: Does this affect the changes that Richard made in his previous changes?

File: commands/priority_test.go [hunk 2]
```diff
@@ -61,7 +61,7 @@ func Test_getOverrideNamespace(t *testing.T) {
 		{
 			stack:    "",
 			flag:     "",
-			want:     "",
+			want:     "openfaas-fn",
```
Comment (alexellis) [inline commands/priority_test.go:64]: I am fairly sure that this will revert the changes by Richard.

**Comment by alexellis** [comment]
For the generateCmd, perhaps change the default value from the const to "openfaas-fn" (in-line) and use a new local variable instead of `functionNamespace` which is used in the other commands. Then revert the change in the current commit.

**Comment by yankeexe** [comment]
> For the generateCmd, perhaps change the default value from the const to "openfaas-fn" (in-line) and use a new local variable instead of `functionNamespace` which is used in the other commands. Then revert the change in the current commit.

@alexellis  Can you check now? I have introduced a new variable (`crdFunctionNamespace`) for namespace flag that deals with just `generate` sub command with default namespace value of `openfaas-fn`.

**Comment by alexellis** [comment]
Looks good @yankeexe, did you do the same testing again, after the change?

Is it also possible for users to give an empty namespace, for the case where they do use the default "" empty namespace?

**Comment by yankeexe** [comment]
> Looks good @yankeexe, did you do the same testing again, after the change?
> 
> Is it also possible for users to give an empty namespace, for the case where they do use the default "" empty namespace?

@alexellis yes, everything is working fine now. With empty string as namespace here is the result. 
image

**Comment by alexellis** [comment]
Thank you for confirming 👍 

I've approved the CI job.

**Comment by goncalo-oliveira** [comment]
Apologies for the late comment, but I'm still seeing this on 0.13.13. I've tested with the same workflow I posted above. Same results.

**Comment by alexellis** [comment]
@LucasRoesler and @yankeexe appear to have verified that the behaviour now works as expected.

Can you confirm what inputs you are using, the exact files, or cleansed copies of them along with the exact commands you are using?

**Comment by goncalo-oliveira** [comment]
It's the same as I posted above, but here it is again. It was rewritten, as I tested again from scratch to make sure...


CLI version check

```bash
$ faas-cli version                                                                      
CLI:
 commit:  72816d486cf76c3089b915dfb0b66b85cf096634
 version: 0.13.13
```

Creating a new function
```bash
$ faas-cli new --lang go --handler src hello
2021/08/03 16:21:05 No templates found in current directory.
2021/08/03 16:21:05 Attempting to expand templates from https://github.com/openfaas/templates.git
2021/08/03 16:21:06 Fetched 13 template(s) : [csharp dockerfile go java11 java11-vert-x node node12 node14 php7 python python3 python3-debian ruby] from https://github.com/openfaas/templates.git
Folder: src created.

Function created in folder: src
Stack file written: hello.yml
```

The `generate` command does not include a default namespace

```bash
$ faas-cli generate -f hello.yml
---
apiVersion: openfaas.com/v1
kind: Function
metadata:
  name: hello
spec:
  name: hello
  image: hello:latest
```

Editing the spec and adding it a namespace

```bash
$ nano hello.yml
$ cat hello.yml
version: 1.0
provider:
  name: openfaas
  gateway: http://127.0.0.1:8080
functions:
  hello:
    lang: go
    handler: ./src
    namespace: openfaas-fn
    image: hello:latest
```

The `generate` command still does not include the namespace

```bash
$ faas-cli generate -f hello.yml
---
apiVersion: openfaas.com/v1
kind: Function
metadata:
  name: hello
spec:
  name: hello
  image: hello:latest
```

Using the flag `--namespace` with the command does work. This is the current workaround.

```bash
$ faas-cli generate --namespace openfaas-fn -f hello.yml
---
apiVersion: openfaas.com/v1
kind: Function
metadata:
  name: hello
  namespace: openfaas-fn
spec:
  name: hello
  image: hello:latest
```

Hope this helps.
